### ISSUE TITLE
[gfx950][MLA] FP8 qh16/q1 persistent sparse decode corrupts Stage-1 output nondeterministically

### ISSUE THREAD

**Issue by yhl-amd** [ROCm/aiter#4638]
## Summary

The gfx950 FP8 persistent sparse-MLA decode kernel below produces nondeterministic Stage-1 attention outputs for byte-identical Q/KV/indices/metadata:

```text
mla_a8w8_qh16_qseqlen1_gqaratio16_ps
_ZN5aiter36mla_a8w8_qh16_qseqlen1_gqaratio16_psE
```

This is not normal split-KV reduction drift. A self-contained test repeatedly invokes the Stage-1 kernel with the same tensors and metadata, and compares the output bitwise with the first run / LSE sibling. The test also checks a PyTorch FP32 reference, verifies that Q/KV/indices and all six metadata buffers remain unchanged, and reports whether corruption originates in Stage 1 or the reducer.

In GLM-5.2 MXFP4 TP8 serving this eventually causes severe generation degradation, including malformed/garbled characters during long generation. The standalone reproducer below does not require the model or ATOM.

## Reproducer and full output

The self-contained pytest and complete raw failure log are attached here:

https://gist.github.com/yhl-amd/2b7313d2a65336410e5dbabb36e8b49a

Files:

- `test_mla_qh16_fp8_persistent_determinism.py`
- `topk2048-repeat50000.log`

Copy the test to `op_tests/`, build/import AITER normally, then run on one gfx950 GPU:

```bash
HIP_VISIBLE_DEVICES=0 \
AITER_TEST_ASM_DIR="$PWD/hsa" \
AITER_TEST_REPEATS=50000 \
pytest -q -s op_tests/test_mla_qh16_fp8_persistent_determinism.py
```

The production-size defaults are intentional:

```text
batch/query count: 3824
query heads:       16
query length:      1
Q/KV dtype:        FP8/FP8
QK head dim:       576 (512 NoPE + 64 RoPE)
output dim:        512 BF16
sparse top-k:      2048
context slots:     20208
page size:         1
max splits:        16
seed:              0
```

With 256 persistent threadgroups, 3824 queries make each threadgroup process about 15 work items. Each query has a distinct sparse page-id stream. Small batches or reusing one identical page-id row for every query often miss the failure.

## Environment

```text
GPU:       MI355X / gfx950:sramecc+:xnack-
ROCm:      7.2.4
HIP:       7.2.53211
PyTorch:   2.10.0+rocm7.2.4.git3d3aa833
```

The current AITER main HSACO was tested:

```text
origin/main at test time: 17f24ec6e93f48722a6b4ec8e54738434194f3d6
HSACO introduced by:      877347d85dcb44da429df28112cb044aadd86a81 (#4452)
HSACO SHA256:             00e2bdfab1dad1767478f85b37aa27520179ecd1d7671b32548ae28b20d18287
```

The Python/JIT checkout used for the standalone run was `3d408300e14d9428cfdc35036c7668548171cb41 + #4415`; #4415 only changes top-k dispatch and is not used by this direct Stage-1 test.

### Current main HSACO (`00e2bdfa...`)

The first repeated non-LSE invocation already differed from the LSE sibling oracle:

```text
reset_metadata=False, repeat=0
mismatched BF16 elements: 114551
bad queries: 17
first bad queries: [60, 75, 150, 750, 960, 1080, 1140, 1275, 1320,
                    1635, 1800, 3195, 3480, 3540, 3555, 3600, 3720]
referenced Stage-1 split-logit mismatches: 0
referenced Stage-1 split-LSE mismatches:   0
direct Stage-1 output mismatches:          114551
input mismatches:    q=0, kv=0, kv_indices=0
metadata mismatches: [0, 0, 0, 0, 0, 0]
```

For query 60, individual heads had hundreds of corrupted values, for example:

```text
head 0:  393 / 512 BF16 values differ
head 2:  486 / 512 BF16 values differ
head 13: 494 / 512 BF16 values differ
head 15: 505 / 512 BF16 values differ
```

### Previous qh16 HSACO (`ceccccd0b`, SHA256 `4e2723dd...`)

A separate 50,000-repeat run failed later but in the same way:

```text
reset_metadata=False, repeat=467
mismatched BF16 elements: 91542
bad queries: 19
first bad query: 87
query 87 / head 0: 458 / 512 values differ
query 87 / head 6: 500 / 512 values differ
referenced Stage-1 split-logit mismatches: 0
referenced Stage-1 split-LSE mismatches:   0
direct Stage-1 output mismatches:          91542
input mismatches:    q=0, kv=0, kv_indices=0
metadata mismatches: [0, 0, 0, 0, 0, 0]
```

The oracle's max absolute error versus a direct FP32 attention calculation was `0.009765625`, while the failing output reached `0.0234375` for the first bad query. This is structured Stage-1 corruption, not merely a different floating-point reduction order.

## Isolation results for the ASM team

The failure has been narrowed to the value operand path before/during the value MFMA:

```text
global V
  -> LDS
  -> ds_read_b64_tr_b8 (transpose load into AGPR)
  -> a120:a127 / a136:a143
  -> value MFMA
```

1. A causal probe around the first value MFMA (`v74:v77`) drifted at repeat 26722 when the natural `a120:a127` source was retained, even with the accumulator initialized and the weight operand controlled.

2. For a second value MFMA (`v82:v85`, source `a136:a143`), the following orthogonal A/B was observed after matching the diagnostic instruction spacing:

| AGPR/V source | accumulator | result |
|---|---|---|
| natural | natural | failed at repeat 17653 |
| fixed | natural | failed at repeat 2763 |
| natural | zero | failed at repeat 5227 |
| fixed | zero | passed 50,000 repeats |

3. With both V source and accumulator controlled, natural and fixed packed-FP8 softmax weights each passed 20,000 repeats. This does not point to the softmax-weight chain as the primary source.

4. Input tensors, metadata, page-derived global addresses, `m0`, transpose-read LDS addresses, and ordinary `ds_read_b64` reads of the corresponding LDS regions were stable. The natural transpose-LDS -> AGPR -> MFMA path was not.

5. There is also a concrete wait-count issue in the mirrored score/K path:

```text
0x528c  s_waitcnt lgkmcnt(4)
0x5290  MFMA consumes a40:a47
```

`a44:a47` comes from the sixth DS request. Changing that wait to `lgkmcnt(3)` made the targeted score-hash probe pass 50,000 repeats instead of failing at repeat 11610. However, this change alone did not eliminate the normal output corruption, so the value/AGPR path still needs a source-level fix.

The evidence supports a missing/incorrect dependency or reuse interval around `ds_read_b64_tr_b8` -> AGPR -> value MFMA in the persistent work loop. It does **not** yet distinguish whether the exact source-level defect is the transpose-load completion, AGPR visibility, or source-register reuse timing.

## Why this is different from #4076

#4076 was traced to stale vLLM metadata caching. This reproducer generates metadata once, reuses byte-identical metadata intentionally, verifies every metadata buffer after the failure, and directly calls the AITER Stage-1 kernel. No vLLM/ATOM/model code is involved.

Reducer-only replay with captured Stage-1 tensors is stable; the mismatches first appear in the direct Stage-1 output rows.

## Requested fix / validation

Could the ASM owners inspect the source/build for:

```text
MLA_A8W8_QH16_1TG_4W_16mx1_32nx4_PS.sp3
```

and fix the source-level LDS/AGPR/MFMA dependency scheduling? A hand-patched HSACO is not intended as the fix.

Suggested acceptance checks:

1. The attached standalone test passes 50,000+ repeats on gfx950.
2. Both LSE and non-LSE qh16/q1 persistent kernels match the FP32 reference tolerance.
3. Q/KV/indices/metadata remain unchanged.
4. GLM-5.2 MXFP4 TP8 long-generation accuracy no longer degrades.

Related HSACO updates: #4341 and #4452.

cc @fangche123 @amd-ruitang3 @shengnxu

**Comment by yhl-amd** [comment]
Update: I adapted AITER's existing Triton `unified_attention_sparse_mla` as an
independent FP8 correctness oracle for the exact GLM-5.2 sparse-decode shape in
this issue.

This does **not** bypass or hide the reported ASM failure: ATOM still dispatches
to `mla_a8w8_qh16_qseqlen1_gqaratio16_ps`, and I have not switched the serving
path. The purpose of this experiment is to run the same attention mathematics
and production-sized sparse inputs through an unrelated implementation on the
same gfx950 GPU.

### Triton adaptation

Starting from AITER commit `34ecbcbf4`, I changed the existing sparse MLA Triton
wrapper/kernel to:

1. accept scalar `q_descale` and `kv_descale` for FP8 Q/K/V;
2. apply the same scale semantics as the generic Triton MLA path:
   `qk_factor = softmax_scale * q_descale * kv_descale`, followed by
   `output *= kv_descale`;
3. decouple the top-k compute tile from KV page size (`page_size=1`,
   `TILE_SIZE=64`), avoiding 2048 one-token loop iterations;
4. preserve an empty online-softmax state across an all-`-1` tile and return
   zero for an all-invalid row. The previous implementation changed `M` from
   `-inf` to `0` on an invalid tile, which biases a later tile when all valid
   logits are negative;
5. retain backward compatibility with the existing BF16 wrapper calls.

The complete modified wrapper, kernel, and standalone pytest are here:

https://gist.github.com/yhl-amd/4daaa2b0d322b47680408bb4b329bc52

### Validation environment and shape

```text
GPU:       gfx950
PyTorch:   2.10.0+rocm7.2.4.git3d3aa833
Triton:    3.7.0
AITER:     34ecbcbf4 plus the Triton-only changes above

Q:         [3824, 16, 576] FP8
KV:        [20208, 1, 1, 576] FP8
top-k:     [3824, 2048] int32
output:    [3824, 16, 512] BF16
```

The production-shape test uses the same RNG ordering as the ASM standalone
reproducer: fixed seed 0, generate FP8 KV, generate FP8 Q, then generate and
sort a distinct random 2048-page row for every query.

The Triton wrapper consumes those physical indices as a fixed 2-D tensor and
does not reuse the ASM work metadata or split-KV reducer. That separation is
intentional: this is a control for the inputs and attention math, not a second
test of the same scheduling implementation.

Results:

```text
50,000 complete production-shape Triton invocations: bitwise identical
Each iteration compares all 31,326,208 BF16 output elements.
Stress-test runtime: 76.56 seconds

FP32 PyTorch reference queries: [0, 60, 1912, 3823]
maximum absolute error: 0.00781250
mean absolute error:    0.00121096
```

Query 60 is included because it was one of the corrupted queries observed in
the ASM reproducer. The Triton result for that input matches the mathematical
reference and does not exhibit the localized Stage-1 corruption.

A smaller test also covers non-unit FP8 descales, page size 1, an invalid
leading 64-token tile, a completely invalid row, and trailing `-1` padding:

```text
maximum absolute error: 0.00341797
mean absolute error:    0.00041115
four invocations:       bitwise identical
```

I also ran the complete pre-existing sparse MLA BF16 test matrix after the
changes:

```text
144 passed in 25.26s
```

Production-shape command:

```bash
AITER_TEST_GLM_PRODUCTION_SHAPE=1 AITER_TEST_REPEATS=50000 \
PYTHONPATH=/path/to/aiter pytest -q \
  op_tests/triton_tests/attention/test_unified_attention_sparse_mla_fp8_glm.py -s
```

### Interpretation

The same FP8 tensors, sparse physical page-id layout, qh16/q1 dimensions, and
production batch/top-k size pass a 50,000-repeat full-output bitwise stress test
through the Triton implementation on gfx950. This matches the repeat window of
the archived ASM failure run (which failed at repeat 467) and strengthens the
conclusion that the nondeterministic corruption reported here is specific to
the persistent ASM Stage-1 kernel/path, rather than an invalid sparse-index
layout or an inherent
FP8/split-KV numerical-accuracy limitation.

This Triton version is currently a correctness oracle only. I have not wired it
into ATOM dispatch or evaluated serving latency, CUDAGraph capture, or TP8
throughput.
